Allyhawkins/setup diffexp data

allyhawkins opened the pull request:
Closes #571 

Here I did some set up to choose our dataset for differential expression. I chose to use the integrated object from the RMS data after integrating with fastMNN and actually started with the combined and integrated datasets directly. The reason for this is because a lot of the setup for that object is repeating the same steps that we are doing in integration and things we have already done, so I wanted to first ensure that this is a dataset that we want to work with. Then in #571 I plan to add in a separate script that will actually create the integrated object from the filtered objects available on ScPCA. The combined and integrated object that I worked with is available in the data integration bucket on S3 for now. 

Some things to keep in mind when reviewing this: 

- The first section is really mostly exploration of the integrated dataset. The integrated dataset I'm starting with consists of 10 RMS samples, but I did some analysis to look at what types of cells are in each of these samples and ended up choosing only 6 to actually move forward with that have similar cell types - 3 that are embryonal RMS and three that are alveolar RMS. If we are good with that, I would only include those in the construction of the integrated dataset. 
- For differential expression, I first started with performing pseudobulking as we will need to do that regardless but then did DGE with both [edgeR ](http://bioconductor.org/books/3.15/OSCA.multisample/multi-sample-comparisons.html)and [DESeq2](https://hbctraining.github.io/scRNA-seq/lessons/pseudobulk_DESeq2_scrnaseq.html) following each of the tutorials. In doing so, I also showed the QC plots that each of them suggested, but I'm a little iffy on them mostly the DESeq2 dataset. The estimated dispersions don't quite show the expected pattern, but in some searching it looks like that may be due to underlying batch effects. I'm curious if others have thoughts or opinions on the two methods. I also did a quick comparison between them and noted that `edgeR` found less significant genes than `DESeq2` did and appears to be more conservative. 
- I did include some feature UMAPs across each sample to validate findings that genes were up/ down regulated by subdiagnosis and I did see some nice differences. Generally I feel pretty confident in being able to use this dataset for differential expression, although as we get into instruction there is a lot more to flush out. 

I attached a copy of the rendered notebook with plots for reference: 
[intro-diff-expression-setup.html.zip](https://github.com/AlexsLemonade/training-modules/files/10167392/intro-diff-expression-setup.html.zip)

sjspielman commented on scRNA-seq-advanced/setup/intro-diff-expression-setup.Rmd:
```diff
@@ -0,0 +1,624 @@
+- Collapses gene expression counts by sample, reflecting that each sample is a replicate rather than each cell. 
+- Masks variance within a sample and only accounts for if a gene is uniformly high or low across samples. 
+
+Let's start with the simple comparison of looking at one cell type and comparing across ERMS and ARMS subtypes. 
```
I think for the actual notebook, we want to be sure to always frame these DE tests from a specific question/hypothesis standpoint instead of more exploratory vibes (which are appropriate here)!

sjspielman commented on scRNA-seq-advanced/setup/intro-diff-expression-setup.Rmd:
```diff
@@ -0,0 +1,624 @@
+                                     group = tumor_myoblast_sce$subdiagnosis)
+tumor_myoblast_sce <- tumor_myoblast_sce[genes_filter,]
+
+# looks like this doesn't actually filter anything though, but we probably want to keep this step in? it seems important 
```
This may be a good teachable to talk about the function defaults and whether they are reasonable for our libraries. Relevant defaults are - 
 - `min.count = 10`: Minimum count required for at least some samples.
 - `min.total.count = 15`: Minimum total count required.
 - `large.n = 10`: Number of samples per group that is considered to be “large”. 
 - `min.prop = 0.7`: Minimum proportion of samples in the smallest group that express the gene.

sjspielman commented on scRNA-seq-advanced/setup/intro-diff-expression-setup.Rmd:
```diff
@@ -0,0 +1,624 @@
+  ),
+  diagnosis_groups = as.factor(diagnosis_groups)) 
+
+# set the levels
+coldata_df$diagnosis_groups <- relevel(coldata_df$diagnosis_groups, ref = "ERMS")
```
I think you need to do this also for edgeR (or remove it so ARMS is always first level), since based on the ggplot compared their significance in the last section the results are opposite sign. So, we want to make sure the direction of comparison is the same if we want to use both DGE methods.

allyhawkins commented on scRNA-seq-advanced/setup/intro-diff-expression-setup.Rmd:
```diff
@@ -0,0 +1,624 @@
+# set up file paths and read in combined SCE and integrated SCE 
+data_dir <- file.path("..", "setup", "SCPCP000005-samples")
+
+combined_sce_file <- file.path(data_dir, "merged_sce", "SCPCP000005_merged_sce.rds")
+combined_sce <- readr::read_rds(combined_sce_file)
+
+
+integrated_sce_file <- file.path(data_dir, "integrated_sce", "SCPCP000005_integrated_fastmnn_sce.rds")
+integrated_sce <- readr::read_rds(integrated_sce_file)
```
Yes they are different! In the setup script we would pull the specific samples used here prior to integration. And explain in the notebook that we are using a slightly different set of samples from the same project.

allyhawkins commented on scRNA-seq-advanced/setup/intro-diff-expression-setup.Rmd:
````diff
@@ -0,0 +1,624 @@
+
+```{r}
+# filter by number of cells 
+low_cell_groups <- tumor_myoblast_sce$ncells < 10
````
Yes that's correct.

allyhawkins commented on scRNA-seq-advanced/setup/intro-diff-expression-setup.Rmd:
```diff
@@ -0,0 +1,624 @@
+                    samples = colData(tumor_myoblast_sce))
+
+# make column names equal to batch ids, helpful for plotting later 
+colnames(dge_list) <- dge_list$samples$batch
```
Yes absolutely. This is something that also needs to be changed from the beginning when setting up the dataset. The batch column should be labeled with `sample` which will change how it's referred throughout teaching.

sjspielman commented on scRNA-seq-advanced/setup/intro-diff-expression-setup.Rmd:
```diff
@@ -0,0 +1,624 @@
+                       colour_by = "celltype_broad",
+                       point_size= 0.5, 
+                       point_alpha = 0.4) +
+  facet_wrap(~ combined_sce$batch, nrow = 3)
```
I did not know you could facet like that with just a variable! Exciting day.

jashapiro replied:
I found a slightly slicker (arguably) way to do this! If you include the `other_fields` argument to scater plot functions, you can use them in ggplot additions. I think it looks nicer, anyway. (You can do more than one as a vector, but I just did one here)
```
                       point_size= 0.5, 
                       point_alpha = 0.4,
                       other_fields = "batch"
                        ) +
  facet_wrap(batch, nrow = 3)

```

allyhawkins commented:
I've made a few edits based on comments from @sjspielman, including: 

- Updating the DESeq analysis to be in the same direction as edgeR (ERMS vs. ARMS) 
- Adding some information about the defaults used for filtering genes in case we want to explore any of them in instruction or talk about any of them. I added the information from Stephanie's comments for our reference. 
- Fixed the wording and typos that were requested. 

This is failing checks at the download data from S3 step... It doesn't seem to be able to verify credentials https://github.com/AlexsLemonade/training-modules/actions/runs/3643152616/jobs/6151067399#step:5:11

jashapiro commented:
> This is failing checks at the download data from S3 step... It doesn't seem to be able to verify credentials https://github.com/AlexsLemonade/training-modules/actions/runs/3643152616/jobs/6151067399#step:5:11

Yes, this is an S3 permissions issue I created and I am trying to solve without reverting completely... It will hopefully be fixed soon!

jashapiro commented:
> > This is failing checks at the download data from S3 step... It doesn't seem to be able to verify credentials https://github.com/AlexsLemonade/training-modules/actions/runs/3643152616/jobs/6151067399#step:5:11
> 
> Yes, this is an S3 permissions issue I created and I am trying to solve without reverting completely... It will hopefully be fixed soon!

I think this is fixed now. I'm also filing a small update to hopefully prevent running this for setup notebooks.

jashapiro reviewed (commented):
Thank you for tackling all this! This looks like a very good start. 

I'll begin by addressing your overall conclusions:
> - The RMS dataset seems like a good dataset we can use for integration

It isn't perfect, but we won't find that.

> - We probably don't want to blindly use all libraries that we have currently used for integration, but should use the 6 samples that have varying sub diagnosis and also share similar cell types

Agreed... One question I do have is whether we want to do 2 different cell types? The exercise might be to do a different cell type, but I am curious if we would get the same DE genes or not, which might be worth a quick look in this notebook. I also asked about whether we get different results if we treat NA cells differently... which might be another question/exploration we could leave for an exercise. 

> - Here we did some preliminary comparisons using `edgeR` and `DESeq`, but we probably don't need to actually do that in class, we should pick one and show how you would do the analysis first to look at one cell type between experimental conditions and then across multiple cell types/ labels
> - `edgeR` is a bit more conservative in its approach to identify differentially expressed genes

My inclination is to stick with `DESeq`, only because that is what we use in bulk, so we have more material that people can go back to and look at. I am not too concerned about differences in how conservative they are... we will just want to again emphasize that interpretation is not binary.

As for looking across multiple cell types, were you thinking you would do more than one in the instruction notebook? Restating what I just said, we might want to leave those for an exercise?

My other big question is about cell type composition comparisons. I am not sure those are appropriate for this dataset, but I do think we want to show how to do that.  Comparing the different tumor cell types here might be interesting, though I don't know if we have the power to see much of anything.

sjspielman reviewed (commented):
At this point, I'd recommend renaming `setup/SCPCP00005-samples/` --> `setup/rms-dge/` or similar, which matches the `rms-integration/` setup directory I have going in #580. That said, I think your `rms-integration/` directory is (and should be!) a different directory from _my_ `rms-integration/`. Even though that metadata TSV is about integration, it is for the DGE instruction module, so I think it's fine to place all your stuff in one DGE `setup/` directory.

allyhawkins commented on scRNA-seq-advanced/setup/intro-diff-expression-setup.Rmd:
````diff
@@ -0,0 +1,633 @@
+Let's do the same thing, but try with DESeq2.
+
+```{r}
+# deseq2 requires the coldata object as a separate input
+coldata_df <- as.data.frame(colData(tumor_myoblast_sce)) %>%
+  dplyr::mutate(diagnosis_groups = dplyr::case_when(
+    subdiagnosis == "Alveolar rhabdomyosarcoma" ~ "ARMS",
+    subdiagnosis == "Embryonal rhabdomyosarcoma" ~ "ERMS"
+  ),
+  diagnosis_groups = as.factor(diagnosis_groups)) 
+
+# set the levels
+coldata_df$diagnosis_groups <- relevel(coldata_df$diagnosis_groups, ref = "ARMS")
````
Just noting that when I didn't have levels, DESeq complained about it which is why I set them in the first place.

allyhawkins commented:
> Agreed... One question I do have is whether we want to do 2 different cell types? The exercise might be to do a different cell type, but I am curious if we would get the same DE genes or not, which might be worth a quick look in this notebook. I also asked about whether we get different results if we treat NA cells differently... which might be another question/exploration we could leave for an exercise.

I agree that looking at one other cell type in this notebook would be worth it to see if we are just seeing global ARMS vs. ERMs differences or cell type specific differences. I did think it would be worthwhile to talk about how you would apply the analysis to all cell types in your dataset, but I wasn't sure if that was something we would want in the instruction notebook or save additional cell types for exercises. For now I'll add one more to this notebook and then we can tweak what we include when we move to instruction? 

> My inclination is to stick with DESeq, only because that is what we use in bulk, so we have more material that people can go back to and look at. I am not too concerned about differences in how conservative they are... we will just want to again emphasize that interpretation is not binary.

I also am on board with sticking with DESeq. Should I leave the edgeR stuff in the notebook so we have it but note that we will use only one method, DESeq, for instruction? 

> My other big question is about cell type composition comparisons. I am not sure those are appropriate for this dataset, but I do think we want to show how to do that. Comparing the different tumor cell types here might be interesting, though I don't know if we have the power to see much of anything.

This is an interesting question and I think it would be great if we can do something like this. I think we may have better luck using the `celltype_fine` labels. I could play around with it? It looks like the section in OSCA that covers this uses edgeR to do this. http://bioconductor.org/books/3.16/OSCA.multisample/differential-abundance.html

allyhawkins commented on scRNA-seq-advanced/setup/intro-diff-expression-setup.Rmd:
````diff
@@ -0,0 +1,633 @@
+
+```{r}
+# put the integrated PCA and UMAP into the combined object so all the data is together 
+reducedDim(combined_sce, "fastmnn_PCA") <- reducedDim(integrated_sce, "fastmnn_PCA")
+reducedDim(combined_sce, "fastmnn_UMAP") <- reducedDim(integrated_sce, "fastmnn_UMAP")
````
I can change it in the one that's being added to the `combined_sce`, but the `integrated_sce` that I'm using we set to have `fastmnn` with lower case from the integration workflow. It will be changed when we use the setup snakemake workflow to create the object, but for now I'm going to leave it.

allyhawkins commented on scRNA-seq-advanced/setup/intro-diff-expression-setup.Rmd:
````diff
@@ -0,0 +1,633 @@
+  dplyr::group_split(batch)
+```
+
+`Tumor_Myoblast` is definitely the top represented cell type across the libraries. 
+So we want to look at myoblasts across our libraries and see if there are any changes in the myoblast population between samples of different types. 
+To do this, we should probably look at our sample metadata to see what types of samples we are looking at. 
````
I like the idea of using different types of pseudobulking in the exercise! And looking at the finer cell types.

allyhawkins commented on scRNA-seq-advanced/setup/intro-diff-expression-setup.Rmd:
````diff
@@ -0,0 +1,633 @@
+
+```{r}
+# filter by number of cells 
+tumor_myoblast_sce <- tumor_myoblast_sce[, tumor_myoblast_sce$ncells >= 10]
````
No there aren't any, but yes we want to show it. Also we could move it to before we filter to just the myoblast population since the filtering we should do across the whole pseudobulked sce.

allyhawkins commented on scRNA-seq-advanced/setup/intro-diff-expression-setup.Rmd:
````diff
@@ -0,0 +1,633 @@
+```
+
+Before we do any DGE, let's do some QC checks to make sure our data looks okay. 
+We make some Mean difference plots, but I'm not really convinced we get a whole lot from these? 
````
The OSCA book shows these individually for each sample? http://bioconductor.org/books/3.16/OSCA.multisample/multi-sample-comparisons.html#performing-the-de-analysis

jashapiro commented:
> I also am on board with sticking with DESeq. Should I leave the edgeR stuff in the notebook so we have it but note that we will use only one method, DESeq, for instruction?
> 
Yes, leave it in this notebook (we might want to rename the notebook "exploration" to better fit its contents)

> > My other big question is about cell type composition comparisons. I am not sure those are appropriate for this dataset, but I do think we want to show how to do that. Comparing the different tumor cell types here might be interesting, though I don't know if we have the power to see much of anything.
> 
> This is an interesting question and I think it would be great if we can do something like this. I think we may have better luck using the `celltype_fine` labels. I could play around with it? It looks like the section in OSCA that covers this uses edgeR to do this. http://bioconductor.org/books/3.16/OSCA.multisample/differential-abundance.html

I am not 100% sold on edgeR for this analysis (no specific reason, just vibes), but it seems like as good a place to start as any... I think we will probably want to be very cautious about recommendations here, and maybe present it as a _thing you might want to look at_ rather than a _here's how to do it_.
On the other hand, I really liked this little section: http://bioconductor.org/books/3.16/OSCA.multisample/differential-abundance.html#de-da-duality

allyhawkins commented on scRNA-seq-advanced/setup/intro-diff-expression-setup.Rmd:
````diff
@@ -0,0 +1,633 @@
+```{r}
+# we get a warning if we don't use local that it can't find a model
+# this could likely be due to batch effects present in the data
+normalized_object <- DESeq2::rlog(deseq_object, blind = TRUE, fitType = 'local')
````
I just tried that and I get the same error message 

> -- note: fitType='parametric', but the dispersion trend was not well captured by the
   function: y = a/x + b, and a local regression fit was automatically substituted.
   specify fitType='local' or 'mean' to avoid this message next time.

allyhawkins commented:
Thanks all for your feedback. I went ahead and incorporated all of the smaller comments that you had. There were some comments about explaining things further that seemed more like notes for expanding in instruction so I didn't directly address those. If you think more explanation at points is needed please let me know. 

A few of the more major changes: 

- Based on one of Josh's comments, I took out the correlation heat map that I had for DESeq. I would prefer to spend more time talking about the results and showing different cell types and didn't think that plot was that informative. 
- I added in some analysis of looking at DE in other cell types. When I did this, I was cringing at copying and pasting and changing variable names, so I wrote a function to perform DESeq and applied it to a set of cell types (specifically the tumor cell types). Then combined all of those results into one dataframe. I was struggling with different ways to represent comparing genes that are differentially expressed in different cell types (that's a whole other can of worms that we can dive into if we want, but seemed like overkill for the stage this is in). Generally, there are genes that are the same in all cell types and are up/down in ARMS or ERMS and then there are some that are cell type specific. I made an example plot to show how you could look at a specific gene across multiple cell types and conditions. I think we could expand on this either in instruction or an exercise. 
- I also did some initial analysis at looking at differential abundance across conditions. I think my take is that I don't know the best way to do this and trust the results. The way I did it here was by getting the counts for each cell type and then comparing them across samples using edgeR. We could do the same thing with DESeq I think. But I don't know if that really is the _best_ way to do it, but also I don't know if there really is a best way yet... I went back to the original paper and looked at the findings they reported as a reference. They found that mesoderm cells were up in ERMS and myocytes were up in ARMS. The fold changes that we see here align with that finding, but the pvalues do not. They also show this on a per sample level in their supplemental figures and each sample is super variable in regards to composition so who really knows... 

Either way here is the updated report. Still some unanswered questions, but I don't think we ever will have a definitive answer for some of them. 
[intro-diff-expression-setup.html.zip](https://github.com/AlexsLemonade/training-modules/files/10198073/intro-diff-expression-setup.html.zip)

sjspielman commented on scRNA-seq-advanced/setup/intro-diff-expression-exploration.Rmd:
````diff
@@ -0,0 +1,800 @@
+
+```{r}
+# set up file paths and read in combined SCE and integrated SCE 
+data_dir <- file.path("..", "setup", "rms")
````
Shouldn't this just be -
```suggestion
data_dir <- file.path("rms")
```
since this is already in `setup/`?

sjspielman commented on scRNA-seq-advanced/setup/intro-diff-expression-exploration.Rmd:
```diff
@@ -0,0 +1,800 @@
+  "SCPCL000479",
+  "SCPCL000480",
+  "SCPCL000481",
+  "SCPCL000484",
+  "SCPCL000488",
+  "SCPCL000491"
```
Just noting that these are NOT also in the integration instruction notebook. Might want to add a small comment about that for future reference.

sjspielman commented on scRNA-seq-advanced/setup/intro-diff-expression-exploration.Rmd:
````diff
@@ -0,0 +1,800 @@
+```
+
+
+## Differential expression of a secondary cell type 
````
```suggestion
## Differential expression of a second cell type 
```

"secondary" sounds to me like we're looking at nested/hierarchical cell types instead of just "a different cell type."

sjspielman commented on scRNA-seq-advanced/setup/intro-diff-expression-exploration.Rmd:
```diff
@@ -0,0 +1,800 @@
+## Some conclusions and future thoughts
+
+- The RMS dataset seems like a good dataset we can use for differential expression analysis
+- We probably don't want to blindly use all libraries that we have currently used for integration, but should use the 6 samples that have varying sub diagnosis and also share similar cell types
```
```suggestion
- We probably don't want to blindly use all libraries that we have currently used for integration, but should use the 6 samples that have varying sub diagnoses and also share similar cell types
```

sjspielman reviewed (approved):
I think this is all set for what we really _need_ for the setup stage! Just a couple minor comments.

jashapiro commented:
> * I also did some initial analysis at looking at differential abundance across conditions. I think my take is that I don't know the best way to do this and trust the results. The way I did it here was by getting the counts for each cell type and then comparing them across samples using edgeR. We could do the same thing with DESeq I think. But I don't know if that really is the _best_ way to do it, but also I don't know if there really is a best way yet... I went back to the original paper and looked at the findings they reported as a reference. They found that mesoderm cells were up in ERMS and myocytes were up in ARMS. The fold changes that we see here align with that finding, but the pvalues do not. They also show this on a per sample level in their supplemental figures and each sample is super variable in regards to composition so who really knows...

I think we can come back to this. My suspicion is that a simpler test statistic could end up significant, but I don't know what test they did. What you have here is very good for a first look!
